[addons] change visualization to new way

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This commit change the visualization addon to the new C++
based addon style.

Also is the visualization process optimized and much more
safe (found during test of old way a lot of faults and crashes).

For this need every visualization addon a update to new style.

Further are parts removed who was never from Kodi and Addon used.

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -19,76 +19,129 @@
+    switch (action.GetID())
+    {
+    case ACTION_VIS_PRESET_NEXT:
+      m_instance->OnAction(VIS_ACTION_NEXT_PRESET, nullptr);
```
missing breaks

AlwinEsch replied:
Upss, becomes changed. Always becomes something wrong during move of agile to here.
Sorry!

ksooo replied:
This is exactly what I ment some time ago when I was saying that the speed of kodi-agile development will not pay out, because at the end there is double effort to port changes to master. :-/

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+    return;
+
+  // FIXME: iAudioDataLength should never be less than 0
+  if (iAudioDataLength < 0)
```
lengths should be unsigned

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+  // Save our audio data in the buffers
+  std::unique_ptr<CAudioBuffer> pBuffer (new CAudioBuffer(iAudioDataLength));
+  pBuffer->Set(pAudioData, iAudioDataLength);
+  m_vecBuffers.push_back(pBuffer.release());
```
m_vecBuffers can be made a vector of `std::unique_ptr` and then you can call `m_vecBuffers.emplace_back(std::move(pBuffer));`

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+  pBuffer->Set(pAudioData, iAudioDataLength);
+  m_vecBuffers.push_back(pBuffer.release());
+
+  if (static_cast<int>(m_vecBuffers.size()) < m_numBuffers)
```
m_numBuffers should be unsigned

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+  if (static_cast<int>(m_vecBuffers.size()) < m_numBuffers)
+    return;
+
+  std::unique_ptr<CAudioBuffer> ptrAudioBuffer(m_vecBuffers.front());
```
`std::unique_ptr<CAudioBuffer> ptrAudioBuffer = std::move(m_vecBuffers.front());` after converting to a vector of unique_ptr

AlwinEsch replied:
Good idea, I change them. Was a c&p from the old :)

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+  if (y + h > g_graphicsContext.GetHeight())
+    h = g_graphicsContext.GetHeight() - y;
+
+  m_instance = new ADDON::CVisualization(std::dynamic_pointer_cast<ADDON::CAddonDll>(addonInfo), x, y, w, h);
```
`static_ptr_cast` to avoid passing an empty pointer

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+   * audio engine stopped before gui!
+   */
+  if (!g_application.m_bStop)
+    CServiceBroker::GetActiveAE().UnregisterAudioCallback(this);
```
This might race. Make ActiveAE always exist like this: https://github.com/xbmc/xbmc/blob/master/xbmc/ServiceManager.cpp#L40

That way, `UnregisterAudioCallback(this)` will just have no effect after ActiveAE is deinitialized

AlwinEsch replied:
I look what possible, but not want to include a change for ActiveAE on this request.
@FernetMenta how you see it to change ActiveAE to always exists?

garbear commented on xbmc/guilib/GUIVisualisationControl.cpp:
```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+    CAudioBuffer* pAudioBuffer = m_vecBuffers.front();
+    delete pAudioBuffer;
+    m_vecBuffers.pop_front();
+  }
```
Once switch to a vector of unique_ptr, this whole block can just become `m_vecBuffers.clear()`

akva2 commented on xbmc/addons/Visualization.cpp:
```diff
@@ -0,0 +1,165 @@
+void CVisualization::transfer_preset(void* kodiInstance, const char* preset)
+{
+  CVisualization *addon = static_cast<CVisualization*>(kodiInstance);
+  if (!addon)
```
preset is user supplied data. needs a nullcheck.

AlwinEsch replied:
You are right becomes added.

akva2 commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Visualization.h:
```diff
@@ -0,0 +1,285 @@
+        m_presetLockedByUser(false)
+    {
+      if (CAddonBase::m_interface->globalSingleInstance != nullptr)
+        throw std::logic_error("kodi::addon::CInstanceVisualization: Creation of more as one in single instance way is not allowed!");
```
.. Cannot create multiple instances of add-on.

akva2 commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Visualization.h:
```diff
@@ -0,0 +1,285 @@
+    void SetAddonStruct(KODI_HANDLE instance)
+    {
+      if (instance == nullptr)
+        throw std::logic_error("kodi::addon::CInstanceVisualization: Creation with empty addon structure not allowed, table must be given from Kodi!");
```
..: Null pointer instance passed.

AlwinEsch committed:
[addons] change visualization to new way

This commit change the visualization addon to the new C++
based addon style.

Also is the visualization process optimized and much more
safe (found during test of old way a lot of faults and crashes).

For this need every visualization addon a update to new style.

Further are parts removed who was never from Kodi and Addon used.

AlwinEsch committed:
[addons] change way to get settings

Due to last changes is it improved by separate the functions for the
types.

AlwinEsch commented:
Ok to bring in? Have mostly on all adons where I can test (linux) changed to new style.

And could the windows addons changed from one who have experience with the OS?

AlwinEsch commented:
Ah, forgot to say. On few I changed also the windows code (where addon also support Linux) but complete not tested here.

Rechi commented:
Which addons are missing?
I think it isn't a big issue if some addons are broken temporary.

AlwinEsch commented:
This six with request (but where also for windows not tested for there):
- visualization.shadertoy
- visualization.projectm
- visualization.fishbmc
- visualization.goom
- visualization.spectrum
- visualization.waveform

This three not:
- visualization.milkdrop
- visualization.milkdrop2
- visualization.vortex (Maybe remove complete until it is stable, has seen during test with old and new way that not stable)

AlwinEsch commented:
jenkins buid this please

AlwinEsch commented:
@Rechi how it looks from your side, OK to bring in?

Rechi commented:
Not looked at the changes, but I can have a look at the mentioned addons and try to adapt them, once this is in.

AlwinEsch commented:
jenkins buid this please

AlwinEsch commented:
@Rechi thanks a lot! If you can look about the addons for Windows, is great.
